packages/nuxt/src/components/module.ts

```diff
@@ -72,7 +73,7 @@ export default defineNuxtModule<ComponentsOptions>({
     nuxt.hook('app:resolve', async () => {
       // components/ dirs from all layers
       const allDirs = nuxt.options._layers
-        .map(layer => normalizeDirs(layer.config.components, layer.config.srcDir))
+        .map(layer => normalizeDirs(layer.config.components, layer.config.srcDir, { priority: layer.config.srcDir === nuxt.options.srcDir ? 1 : 0 }))
```

Comment:
As another idea, we could give main project 0 and layer -1, -2, -3 etc (it also doesn't requires reverse i guess)